daos: daos_mknod implementation

daltonbohning opened the pull request:
daos_mknod is implemented in two steps:
1. dfs_open(..., O_CREATE,...)
2. dfs_release(...)

This works for regular files only at this time.

closes #377 
closes #353 

Signed-off-by: Dalton Bohning <daltonx.bohning@intel.com>

daltonbohning commented:
closes #377

adammoody commented:
Thanks @daltonbohning .  You're really knocking out a lot of items over this past week!  Nice!

It's fine to just support regular files for now, especially since the tools don't generally support types other than directories, regular files, and symlinks just yet.  However, let's add a check and print a warning message to alert the user if this mknod function happens to get called with something other than a regular file.  In that case, perhaps we could also set errno and return -1 if there is good errno value that matches up.

Also, it'd be helpful for future developers to add a comment in the function just to mention that the goal is to emulate a mknod operation using an open/release.

adammoody commented:
If printing/returning errors would cause problems, like lots of harmless warnings under typical usage, we can reconsider.  I think the main reason we call mknod right now is to set lustre striping on files, so it's kind of a special case.  We could double check the rest of mpiFileUtils code to be sure.

daltonbohning commented:
I revised my comment to be more clear, after some further investigation into dfs_open.
dfs_open allows regular files, symbolic links, and directories. If anything else is supplied, dfs_open returns an appropriate error, which daos_mknod captures and propogates:
https://github.com/daos-stack/daos/blob/73b42ec7870d74091a074e2b189f8d2d48f4fd53/src/client/dfs/dfs.c#L2315-L2340

daltonbohning commented:
As for mpifileutils calling mknod, I found this was necessary for DAOS (and possible regular POSIX as well, though I'm not sure).
I ran into an issue where if multiple processes were writing different chunks to the same file, and that file did not yet exist, then the destination file would _sometimes_ not contain all of the written chunks. But if the destination file did exist before copying any chunks, then there were no issues.
mfu_flist_copy calls mfu_file_create, which calls mfu_file_mknod BEFORE copying any data. And since mknod creates the file, it solves this issue.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -449,10 +449,50 @@ int mfu_lstat64(const char* path, struct stat64* buf)
+    if (parent != NULL) {
+        /* create regular file */
+        rc = dfs_open(mfu_file->dfs, parent, name,
+                      S_IFREG | mode, O_CREAT,
```
I know @dsikich mentioned adding ``S_IFREG`` was needed here but now I forget why.  Can you also add a comment about that here so I don't forget again?

dsikich replied:
@adammoody if you don't specify S_IFREG then dfs_open will return with an error saying "invalid inputs." It expects you to tell it if it is a regular file.

daltonbohning replied:
Thanks @dsikich 
I'll check on this for my own information and leave an appropriate comment.

dsikich replied:
@daltonbohning yeah if you go tohttps://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c in dfs_open it has a switch statement that checks the file type. If it doesn't match then it returns EINVAL, maybe there is a way around passing it in explicitly, not %100 sure on that.

daltonbohning replied:
(Please let me know if any of this doesn't make sense)

So the call to `mfu_file_mknod` within `mfu_flist_copy` actually passes in `S_IFREG` directly:
https://github.com/daltonbohning/mpifileutils/blob/cb2d35c20662569dc1385f375fe17e267b898b90/src/common/mfu_flist_copy.c#L1130

But the definition for `mfu_file_open` does not have a `mode` parameter directly. Instead, `mode` can be passed to `mfu_file_open` through "variable arguments", and not all calls to `mfu_file_open` pass a `mode`, and hence do not pass `S_IFREG`, which is why `daos_open` uses `S_IFREG | mode`, in case it was not passed in.
https://github.com/daltonbohning/mpifileutils/blob/cb2d35c20662569dc1385f375fe17e267b898b90/src/common/mfu_io.c#L706

`mfu_open` checks for variable arguments, and passes `mode` to `open` only if `O_CREAT` is specified, as is required by `open`:
https://man7.org/linux/man-pages/man2/open.2.html
If `mode` is not passed to `open`, and the file already exists, it doesn't matter to POSIX, because it will open the file of whatever type it is.

But in DAOS, `dfs_open` DOES require `mode` with the filetype, even if the file already exists.

POSIX `mknod`, however, always requires `mode` to be passed in to specify the file type.

But I wonder if doing `S_IFREG | mode` might be a problem if `mode = S_IFDIR` is passed into `daos_open`?
Since `dfs_open` only supports `S_IFREG`, `S_IFDIR`, `S_IFLNK`, maybe `daos_mknod` and/or `daos_open` should more appropriately have something like this (but in a better-coded way):
```
switch (mode & S_IFMT) {
	case S_IFREG:
	        // good
	case S_IFDIR:
	        // good
	case S_IFLNK:
	        // good
	default:
                set mode = S_IFREG
```
This would allow any of `S_IFREG`, `S_IFDIR`, `S_IFLNK` to be be passed to `daos_open` or `daos_mknod`, and would only default `mode` if neither are passed in.

daltonbohning replied:
Update: `dfs_open` requires `value` to be set if the type is `S_IFLINK`, so maybe we would only allow `S_IFREG`, `S_IFDIR` for now.

dsikich replied:
@daltonbohning not sure why we would ever call dfs_open on a directory? we should be using daos_mkdir, daos_opendir for those operations I would think.

daltonbohning replied:
In that case, I think the `S_IFREG` would then be the only one allowed, as it is now.
And, as Adam suggested previously, I think I'll add a check to make sure that's the only type passed to `mknod`, and leave `daos_open` with the `S_IFREG | mode`, as you originally put in.
(But I'll leave some comments so we don't need this discussion again)

dsikich replied:
@daltonbohning yeah that sounds good

adammoody commented:
> 
> 
> As for mpifileutils calling mknod, I found this was necessary for DAOS (and possible regular POSIX as well, though I'm not sure).
> I ran into an issue where if multiple processes were writing different chunks to the same file, and that file did not yet exist, then the destination file would _sometimes_ not contain all of the written chunks. But if the destination file did exist before copying any chunks, then there were no issues.
> mfu_flist_copy calls mfu_file_create, which calls mfu_file_mknod BEFORE copying any data. And since mknod creates the file, it solves this issue.

Interesting.  Sounds like a race condition on creating the "inode" or something.  Based on your debugging, it sounds like this avoids the problem.  Do you know whether that is expected behavior in DAOS?

daltonbohning commented:
> Interesting. Sounds like a race condition on creating the "inode" or something. Based on your debugging, it sounds like this avoids the problem. Do you know whether that is expected behavior in DAOS?

I checked with regular POSIX files (removing the `mknod` call) and was not able to reproduce the issue.

I'm not sure if it is expected behavior in DAOS, but since POSIX works without issues, it seems `daos_mknod` "indirectly" corrects the issue. I would like to find a more proper fix for this. I'll need to do some further digging. It may be that the `dfs` library just isn't explicitly handling race conditions, since it doesn't use MPI, etc.

dsikich commented:
> > Interesting. Sounds like a race condition on creating the "inode" or something. Based on your debugging, it sounds like this avoids the problem. Do you know whether that is expected behavior in DAOS?
> 
> I checked with regular POSIX files (removing the `mknod` call) and was not able to reproduce the issue.
> 
> I'm not sure if it is expected behavior in DAOS, but since POSIX works without issues, it seems `daos_mknod` "indirectly" corrects the issue. I would like to find a more proper fix for this. I'll need to do some further digging. It may be that the `dfs` library just isn't explicitly handling race conditions, since it doesn't use MPI, etc.

@adammoody @daltonbohning I can't say for sure either. It is a bit puzzling to me since I would think that on the open call in mfu_copy_open_file it would open and create the file if it doesn't exist, then write to its particular offset with the length of bytes it was given.

daltonbohning commented:
> I can't say for sure either. It is a bit puzzling to me since I would think that on the open call in mfu_copy_open_file it would open and create the file if it doesn't exist, then write to its particular offset with the length of bytes it was given.

I'll look into this as well, to rule out `mfu` as being the culprit.

daltonbohning commented:
> las p1cont1 p1cont2; rm p1cont2/a

@adammoody Turns out this was in fact a bug in `dfs.c` that was fixed on Thursday. But we needed the `daos_mknod` implementation anyway, so that's a win!

daltonbohning commented:
> If printing/returning errors would cause problems, like lots of harmless warnings under typical usage, we can reconsider. I think the main reason we call mknod right now is to set lustre striping on files, so it's kind of a special case. We could double check the rest of mpiFileUtils code to be sure.

I think it will be okay to do as you suggest, since `mfu_file->type == DAOS` is really only being used from `dcp.c` at this time, and the DAOS team (Danielle and I) will have to run our tests again if ever mpifileutils is updated, so we should catch any future issues.

daltonbohning commented:
@adammoody I think I've finished with this one now. I added a little more cleanup, and switched an error check in the spirit of "keeping an `rc` check close to the function call", i.e. `if parent == NULL then error` instead of `if parent != NULL... else error`

adammoody commented on src/common/mfu_io.c:
```diff
@@ -609,6 +669,8 @@ void daos_open(const char* file, int flags, mode_t mode, mfu_file_t* mfu_file)
     mfu_free(&name);
     mfu_free(&dir_name);
+
+    return -1;
```
Should this be ``return rc``?

Also, if ``#ifdef DAOS_SUPPORT`` is not active, I'm guessing some compilers will not be happy with this function, since it is declared to return an int, but it wouldn't have a return statement in that case.  To fix that, you could add an ``#else`` case or perhaps but an ``int rc = 0;`` at the top and an ``return rc;`` at the bottom?

adammoody replied:
On second thought, I see the ``#ifdef`` is the same on all daos functions, so let's wait to do that in a future PR if it's actually needed.

daltonbohning replied:
> Should this be `return 0`?
Yes, that was careless of me... I'll need to investigate my tests didn't catch this...
But it has been fixed.

daltonbohning replied:
> `int rc = 0;` at the top and an `return rc;` at the bottom

I think this should be easy enough to do. I can do that when I also go through this one:
https://github.com/hpc/mpifileutils/issues/382
Since I will be touching most functions anyway.

adammoody commented on src/common/mfu_io.c:
```diff
@@ -609,6 +669,8 @@ void daos_open(const char* file, int flags, mode_t mode, mfu_file_t* mfu_file)
 
     mfu_free(&name);
     mfu_free(&dir_name);
```
Oh, we should have to free up these strings under different error cases.

daltonbohning replied:
I think this should be fine, no? The error cases fall through to the `mfu_free`s, except the error case before `name` and `dir_name` are declared.

adammoody replied:
Yes, you're right.  Sorry, I was skimming too quickly and misreading those ``rc = -1`` statements in the errors as ``return -1``.

Good to rebase and merge.

Thanks!

daltonbohning replied:
That's okay :)
Thanks for all the feedback!
